vms/avm/states/state.go

```diff
@@ -576,3 +597,175 @@ func (s *state) writeStatuses() error {
 	}
 	return nil
 }
+
+func (s *state) Prune(lock sync.Locker, log logging.Logger) error {
+	lock.Lock()
+	// It is possible that more txs are added after grabbing this iterator. No
+	// new txs will write a status, so we don't need to check those txs.
+	statusIter := s.statusDB.NewIterator()
+	defer statusIter.Release()
+
+	if !statusIter.Next() {
+		// If there are no statuses on disk, pruning was previously run and
+		// finished.
+		lock.Unlock()
+
+		log.Info("state already pruned")
+
+		return statusIter.Error()
+	}
+
+	startTxIDBytes := statusIter.Key()
+	txIter := s.txDB.NewIteratorWithStart(startTxIDBytes)
+	defer txIter.Release()
+
+	// While we are pruning the disk, we disable caching of the data we are
+	// modifying. Caching is re-enabled when pruning finishes.
+	//
+	// Note: If an unexpected error occurs the caches may never be re-enabled.
+	// That's fine as the node is going to be in an unhealthy state regardless.
+	oldStatusCache := s.statusCache
+	oldTxCache := s.txCache
+	s.statusCache = &cache.Empty[ids.ID, *choices.Status]{}
+	s.txCache = &cache.Empty[ids.ID, *txs.Tx]{}
+	lock.Unlock()
+
+	startTime := time.Now()
+	startProgress := timer.ProgressFromHash(startTxIDBytes)
+
+	startStatusBytes := statusIter.Value()
+	if err := s.cleanupTx(lock, startTxIDBytes, startStatusBytes, txIter); err != nil {
+		return err
+	}
+
+	i := 1
+	for statusIter.Next() {
+		i++
+
+		txIDBytes := statusIter.Key()
+		statusBytes := statusIter.Value()
+		if err := s.cleanupTx(lock, txIDBytes, statusBytes, txIter); err != nil {
+			return err
+		}
+		if i%pruneCommitLimit == 0 {
+			// We must hold the lock during committing to make sure we don't
+			// attempt to commit to disk while a block is concurrently being
+			// accepted.
+			lock.Lock()
+			err := s.Commit()
+			lock.Unlock()
+			if err != nil {
+				return err
+			}
+
+			progress := timer.ProgressFromHash(txIDBytes)
+			eta := timer.EstimateETA(
+				startTime,
+				progress-startProgress,
+				math.MaxUint64-startProgress,
+			)
+			log.Info("committing state pruning",
+				zap.Int("count", i),
+				zap.Duration("eta", eta),
+			)
+		}
```

Comment:
May make sense to put some sleeps in here to avoid choking the DB. I think we did this for some P-Chain iterations in the past?

Reply:
We have done this in the past... (And still have that code in the [proposervm height indexer](https://github.com/ava-labs/avalanchego/blob/eb6e7973a900dc647d29a2da8e7a233ac0a5fe77/vms/proposervm/indexer/height_indexer.go#L197))... I felt like it was a bit overkill here... But I'm fine adding a few ms delay after commit if we wanted to.

Reply:
I would prefer 1024 txs per iteration with 5x sleep.